[mesos/docker] Enhance kube-up to better support running in a container (for CI)

karlkfi opened the pull request:
- Generate CA & API Server SSL key/cert in keygen docker image
  - Refactor SSL generation
  - Generate service account key & user files on local machine
- Enable kube-up to be run in a container (kubernetes-mesos-test)
- Add timeout env vars
  - KUBE_KEYGEN_TIMEOUT (default: 60s)
  - MESOS_DOCKER_ETCD_TIMEOUT (default: 60s)
  - MESOS_DOCKER_MESOS_TIMEOUT (default: 60s)
  - MESOS_DOCKER_API_TIMEOUT (default: 180s)
  - MESOS_DOCKER_ADDON_TIMEOUT (default: 180s)
- Pull docker images up front to avoid timeouts
- Remove docker image builds from test-setup
- Nuke logs dir before each kube-up
- Make run_in_docker work without KUBECONFIG defined
- Fix temp dir cleanup
- Add MESOS_DOCKER_WORK_DIR env var
  - Usage: logs, auth, mesos slave workspaces
  - Default to $HOME/tmp/kubernetes
  - Outside of repo (which gets docker mounted when using kubernetes-mesos-test)
  - Inside $HOME (which gets vm mounted when using docker-machine or boot2docker)
- Add MESOS_DOCKER_WORK_DIR env var
  - Usage: docker-in-docker image storage
  - Default to /var/tmp/kubernetes (survive host reboots)
  - Outside of repo (which gets docker mounted when using kubernetes-mesos-test)
  - Outside $HOME (which gets vm mounted when using docker-machine or boot2docker)
- Enable errtrace
- Increase log level to aid CI debugging
- Improve mesos slave missing error in detect-minions

Fixes `make tests_e2e` run in `mesosphere/kubernetes-mesos-test`.

Mesos-only cluster deployment changes.

xref: https://github.com/mesosphere/kubernetes-mesos/issues/305

Commit:
[mesos/docker] Enhance kube-up to better support running in a container (for CI)

- Generate CA & API Server SSL key/cert in keygen docker image
  - Refactor SSL generation
  - Generate service account key & user files on local machine
- Enable kube-up to be run in a container (kubernetes-mesos-test)
- Add timeout env vars
- Pull docker images up front to avoid timeouts
- Remove docker image builds from test-setup
- Nuke logs dir before each kube-up
- Make run_in_docker work without KUBECONFIG defined
- Fix temp dir cleanup
- Add auth mount env var
  - Default to $HOME/tmp/kubernetes/auth
  - Outside of repo (which gets docker mounted when using kubernetes-mesos-test)
  - Inside $HOME (which gets vm mounted when using docker-machine or boot2docker)
- Add log dump dir env var
  - Default to $HOME/tmp/kubernetes/logs (for consistancy with auth dir)
- Enable errtrace
- Increase log level to aid CI debugging

karlkfi commented:
rebased

karlkfi commented:
Looks like `test-setup` is called after `kube-up.sh` in `e2e-up.sh`. So creating the images there is pointless, since kube-up already builds them.

Somewhere we need to nuke the `cluster/mesos/docker/logs` dir, but since kube-up populates it on ERR, it should probably be done at the beginning of kube-up.

I've also found that pulling the docker images on a fresh machine can cause the `docker-compose up` to timeout. So I'll probably merge these 3 issues into one PR.

karlkfi commented:
Added env vars to allow users (e.g. CI) with systems/networks to adjust timeouts.

karlkfi commented:
Flakey test?

```
Summarizing 1 Failure:

[Fail] Services [It] should serve multiport endpoints from pods 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:1132

Ran 84 of 142 Specs in 384.295 seconds
FAIL! -- 83 Passed | 1 Failed | 2 Pending | 56 Skipped 
```

```
• Failure [118.709 seconds]
Services
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:952
  should serve multiport endpoints from pods [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:235

  Aug 10 20:33:25.638: Timed out waiting for service multi-endpoint-test in namespace e2e-tests-service-0-410m4 to expose endpoints map[podname1:[100] podname2:[101]] (1m0s elapsed)

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:1132
```

karlkfi commented:
Mesosphere CI still doesn't even finish kube-up with these changes:

```
Waiting (up to 300s) for /var/run/kubernetes/auth/apiserver.crt to exist
Timed out
```

jdef commented:
lgtm, though i haven't tested this. is there a process diagram somewhere to explain how everything is chained together? it keeps getting more complicated.. :)

tsenart commented on cluster/mesos/docker/docker-compose.yml:
```diff
@@ -149,9 +156,29 @@ scheduler:
+  volumes:
+  - ${MESOS_DOCKER_VOLUME_ROOT}/auth:/var/run/kubernetes/auth
+  links:
+  - apiserver
```
New line?

tsenart commented on cluster/mesos/docker/sslgen/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+COPY ./bin/* /usr/local/bin/
+
+ENTRYPOINT ["k8sm-sslgen.sh"]
+CMD ["/var/tmp/kubernetes/auth"]
```
New line?

tsenart commented:
LGTM as far as I can tell. Although I'm quite terrified of these massive amounts of shell scripts.
Would we envision a future where we could use Go for these things? For instance, the crypto packages are quite good and I wonder if they'd be able to match up for these requirements.

karlkfi commented:
Having slept on it, I think I get to be my own worst enemy.

Pros:
- `docker-compose up` works with minimal setup
- Generating SSL in a container allows a controllable openssl version
- MESOS_DOCKER_VOLUME_ROOT adds flexibility to work around the annoying fact that docker-machine and boot2docker only share `/Users` with their host.
- Configurable timeouts allow short timeouts for dev usage and longer for CI

Cons:
- Adding two containers that run once and die into a docker-compose config feels wrong
  - Hard to read the flow
  - Confusing to a new user that two are required
  - Wrong tool for the job...
- MESOS_DOCKER_VOLUME_ROOT complexity:
  - `/var/run/kubernetes` (in docker-compose.yml config)
  - `/var/tmp/kubernetes` (default in sslgen)
  - `/tmp/kubernetes` (used by CI)
  - `${HOME}/tmp/kubernetes` (used by devs)
- Too many timeouts to configure
- More docker images to build and ship around

sttts commented on cluster/mesos/docker/docker-compose.yml:
```diff
@@ -110,20 +114,23 @@ controller:
+  - MESOS_DOCKER_API_TIMEOUT
   volumes:
-  - ./certs/controller:/var/run/kubernetes/auth
+  - ${MESOS_DOCKER_VOLUME_ROOT}/auth:/var/run/kubernetes/auth:ro
```
Nice with the new env interpolation in docker-compose.

sttts commented on cluster/mesos/docker/docker-compose.yml:
```diff
@@ -149,9 +156,29 @@ scheduler:
   - mesosslave1
   - mesosslave2
   - apiserver
+sslgen:
```
funny to see this. I had the same approach before moving everything into util.sh.

sttts commented on cluster/mesos/docker/docker-compose.yml:
```diff
@@ -149,9 +156,29 @@ scheduler:
+  entrypoint: [ "k8sm-sslgen.sh" ]
+  command: [ "/var/run/kubernetes/auth" ]
+  volumes:
+  - ${MESOS_DOCKER_VOLUME_ROOT}/auth:/var/run/kubernetes/auth
```
Why no volume mount from this container directly?

sttts replied:
Then we can get rid of the temporary directory altogether

karlkfi replied:
kubectl configuration still needs access to the `root-ca.crt` and `token-users`, but now that I think about it, we can probably use `docker cp` to get them out.

sttts commented on cluster/mesos/docker/util.sh:
```diff
@@ -152,68 +160,54 @@ function verify-prereqs {
+    # Pull before `docker-compose up` to avoid timeouts between container runs.
+    # Pull before building images (but only if built) to avoid overwriting locally built images.
+    echo "Pulling docker images" 1>&2
+    docker-compose -f "${compose_file}" pull
```
Didn't know this is possible. :+1:

karlkfi replied:
Downside is that it pulls whether you have a newer one built locally or not.

karlkfi commented:
rebased to fix shippable error

karlkfi commented:
Squashed.

karlkfi commented:
@zmerlynn You wanna give this bashiness a look-see?

Commit:
[mesos/docker] Configure controller-manager to advertize its externally accessible IP

zmerlynn commented:
Sorry, I didn't see this. Will look soon.

Commit:
[mesos/docker] Replace auth/log dirs with MESOS_DOCKER_WORK_DIR

- Mount mesos slave workspaces in MESOS_DOCKER_WORK_DIR
- Clear mesos slave workspace in kube-up
- Add MESOS_DOCKER_IMAGE_DIR to configure docker-in-docker image storage
- Lazily eval auth/log dirs
- Improve mesos slave missing error in detect-minions

Commit:
Update mesos-slave-dind to use Docker 1.8.1

Commit:
[mesos/docker] Upgrade mesos-docker-test image to use docker 1.8.1

Commit:
[mesos/docker] Expose DOCKER_DAEMON_ARGS

karlkfi commented:
@zmerlynn You gonna have a chance to take a look at this or should I just get @sttts to give it the ok?
It won't affect anything outside of the mesos/docker cluster. I mostly just roped you in to increase the bus factor and get some feedback on consistency with other cluster's bash scripts. The containerized SSL cert generation, for example, might be useful elsewhere.

zmerlynn commented:
Ack, sorry! I'll look today.

karlkfi commented:
Thanks, @zmerlynn !
